### ISSUE TITLE
feat: Add hover information on all Cairo variables, functions, modules etc. in VS Code extension

### ISSUE THREAD

**Issue by misicnenad** [starkware-libs/cairo#5711]
# Feature Request

## Consolidated summary

Implement the following features in hovers system. Each feature is probably a separate task:

- [x] When I hover over a Cairo function, variable, module etc., both from the core lib and from custom modules, a tooltip should appear containing useful information about the underlying item. This information should be read from related comments explaining the item if they exist; otherwise, no info should/can be shown. The comments that can be used for purposes of both documenting code _and_ showing in this on-hover tooltip can be the triple slash ones (`///`).

    Additionally, hovering over items in the `use` statement should display a tooltip too (currently nothing is displayed):
![image](https://github.com/starkware-libs/cairo/assets/19427053/336b0b71-40bc-43c0-a14e-4d35542fc5c3)
- [x] moved to #5823 ~`rust-analyzer` even provides this capability for *.toml* files:
![image](https://github.com/starkware-libs/cairo/assets/19427053/d0f96df0-0333-496c-9324-2dd7cbd6ce1f)~
- [x] moved to #5824 ~If I hover over `println`, I should see something like:
![image](https://github.com/starkware-libs/cairo/assets/19427053/b27dcea0-460e-476d-ad70-aaf30e72bd7c)~
- [x] and when I hover over `foo`, I should see:
![image](https://github.com/starkware-libs/cairo/assets/19427053/c3935cf3-07b2-486e-b4c1-58ddc74a201f)
- [x] When I hover over `core::cmp::max`, I'd expect something like this:
![image](https://github.com/starkware-libs/cairo/assets/19427053/5695a6fc-688e-4ae8-a0d6-727c84815d5f)
- [x] Hovers don't really work as described in the issue, they display just the import path (if a struct is in question) or the function signature, which is insufficient imo:
![image](https://github.com/starkware-libs/cairo/assets/19427053/fef3dfe7-436c-4167-b895-865a609a807d)
- [x] Showing function signatures is great, but showing the import paths for each parameter's type makes it too cluttered.
![image](https://github.com/starkware-libs/cairo/assets/19427053/0175461f-8c52-4a5e-a23c-70d495481fa3)
- [x] Additionally, there is *no* hover information when hovering over items being imported:
![image](https://github.com/starkware-libs/cairo/assets/19427053/b04487cd-e0f9-4f46-8ffc-c91bcbe0a198) What I'd expect is something like this:
![image](https://github.com/starkware-libs/cairo/assets/19427053/55925208-7f3c-453a-bfcb-f74fcfc9d3d3)

---

## Original issue description below

**Describe the Feature Request**

Hovering over any Cairo variable, function, module etc. should display useful information. This information is set by using appropriate comment format, similar to how it's done for Rust and its corresponding VS Code extension `rust-analyzer`.

Example in Rust:
![image](https://user-images.githubusercontent.com/48062697/113020658-b5f98b80-917a-11eb-9f88-3dbc27320c95.gif)

**Describe Preferred Solution**

When I hover over a Cairo function, variable, module etc., both from the core lib and from custom modules, a tooltip should appear containing useful information about the underlying item. This information should be read from related comments explaining the item if they exist; otherwise, no info should/can be shown. The comments that can be used for purposes of both documenting code _and_ showing in this on-hover tooltip can be the triple slash ones (`///`).

Additionally, hovering over items in the `use` statement should display a tooltip too (currently nothing is displayed):
![image](https://github.com/starkware-libs/cairo/assets/19427053/336b0b71-40bc-43c0-a14e-4d35542fc5c3)

**Describe Alternatives**

`rust-analyzer` even provides this capability for *.toml* files:
![image](https://github.com/starkware-libs/cairo/assets/19427053/d0f96df0-0333-496c-9324-2dd7cbd6ce1f)

**Related Code**

Using the following Cairo code:
```rust
use core::cmp::max;

/// Some useful info
fn foo() {}

fn main() {
    foo();
    println!("");
}
```
If I hover over `println`, I should see something like:
![image](https://github.com/starkware-libs/cairo/assets/19427053/b27dcea0-460e-476d-ad70-aaf30e72bd7c)

and when I hover over `foo`, I should see:
![image](https://github.com/starkware-libs/cairo/assets/19427053/c3935cf3-07b2-486e-b4c1-58ddc74a201f)

When I hover over `core::cmp::max`, I'd expect something like this:
![image](https://github.com/starkware-libs/cairo/assets/19427053/5695a6fc-688e-4ae8-a0d6-727c84815d5f)

**Additional Context**

https://rust-analyzer.github.io/manual.html#hover

**If the feature request is approved, would you be willing to submit a PR?**
_(Help can be provided if you need assistance submitting a PR)_

- No

Don't have enough bandwith atm, but will ping someone here once I do.

**Comment by mkaput** [comment]
Few notes:
1. A big chunk of this is already implemented, you can hover over references to functions/structs/consts/other items.
2. The current implementation is unaware of local variables/struct fields and will show hover for encompassing item. I want to add some hacks to improve this (cc @spapinistarkware), but for general fix we plan to wait for `scarb doc` to happen and upstream their much better extraction logic (cc @drknzz, @Arcticae). So that LS could reuse it.
3. Macros are the hardest part here because they are in no way defined in Cairo, instead being built into the running compiler. So the example from description issue will happen in far future if it will at all 😕

**Comment by misicnenad** [comment]
@mkaput 
> A big chunk of this is already implemented...

What are you referring to here, maybe I'm missing something? 

Hovers don't really work as described in the issue, they display just the import path (if a struct is in question) or the function signature, which is insufficient imo:
![image](https://github.com/starkware-libs/cairo/assets/19427053/fef3dfe7-436c-4167-b895-865a609a807d)

Showing function signatures is great, but showing the import paths for each parameter's type makes it too cluttered.
![image](https://github.com/starkware-libs/cairo/assets/19427053/0175461f-8c52-4a5e-a23c-70d495481fa3)


----
Additionally, there is *no* hover information when hovering over items being imported:
![image](https://github.com/starkware-libs/cairo/assets/19427053/b04487cd-e0f9-4f46-8ffc-c91bcbe0a198)

What I'd expect is something like this:
![image](https://github.com/starkware-libs/cairo/assets/19427053/55925208-7f3c-453a-bfcb-f74fcfc9d3d3)

**Comment by mkaput** [comment]
> What are you referring to here, maybe I'm missing something?

Yes, that's what I meant. Thanks for all these examples. I agree things can be improvement, and your screenshots will now function as milestones to achieve here.



> Regarding waiting for `scarb doc`, do you have links at hand to read more about this?

https://github.com/software-mansion/scarb/tree/main/extensions/scarb-doc

The team is in a before-MVP phase and are mostly experimenting. We do not have any written prose about it yet.



> Sad to hear about macros not being supported soon.. Aren't they built similarly to how they're built in Rust? Can't their logic be an inspiration on implementing this in Cairo then?
> I'm probably missing something again.

No, in Cairo macros are written in Rust and are either (currently) built-in/hardcoded in the compiler logic or (in private alpha testing) written in Rust and compiled as shared objects and dynamically loaded by Scarb. There are no plans to implement macros-in-Cairo because this would be not performant at all, inconvenient to write etc.

Another difference is that although macros do [declare what `!` identifiers they **can** consume](https://github.com/starkware-libs/cairo/blob/355055b82626662638dac88810cb9258a2ac9d45/crates/cairo-lang-defs/src/plugin.rs#L87-L114), but actually, they are not obliged to consume what they declare. In Rust, on the other hand, macros are declared like functions, and the compiler is the one that ensures macros are expanded. Cairo macros just get a snippet of source encompassed by the macro and are free to do whatever they want, including leaving the macro as-is. This makes the operation of learning what macro compiler will expand at the given cursor position quite complex.

**Comment by mkaput** [comment]
Adding `help wanted` label because this is a frequently appearing request for improvement, but there are fundamental things that are more important that I have to work back on. If nobody were willing to contribute this, I will eventually get back to this issue in the future.

Mind that most of the cases described by @misicnenad are on the harder side of things implementation-wise and may require smart refactorings in the compiler.

**Comment by mkaput** [comment]
Added all examples at the top of issue description

**Comment by LucasLvy** [comment]
Functions signature are working properly on my machine (the format is a bit broken but you get the informations you're looking for. Can you make sure that you're using the latest scarb released version ?
![Screenshot 2024-06-05 at 10 16 00](https://github.com/starkware-libs/cairo/assets/70894690/0e8d38a4-acf5-4e7c-83d2-86ee3199f5b7)

**Comment by misicnenad** [comment]
@LucasLvy 
I'm not seeing this:
![image](https://github.com/starkware-libs/cairo/assets/19427053/9a78fbdf-3f05-4029-97f7-f9e1eab46221)

Using:
```
scarb 2.6.4 (c4c7c0bac 2024-03-19)
cairo: 2.6.3 (https://crates.io/crates/cairo-lang-compiler/2.6.3)
sierra: 1.5.0
Cairo1.0 (VS Code): 2.6.3
```

**Comment by mkaput** [comment]
Ah yeah. Try nightly please

**Comment by misicnenad** [comment]
Thanks @mkaput . One more thing regarding tooltips, it would be nice to have colors, similar to Rust - it would make it much easier to quickly gather what one is seeing

**Comment by mkaput** [comment]
> Thanks @mkaput . One more thing regarding tooltips, it would be nice to have colors, similar to Rust - it would make it much easier to quickly gather what one is seeing

Yes I know. For this, we need a [VSCode grammar file](https://code.visualstudio.com/api/language-extensions/syntax-highlight-guide) to be written so I also had to put this on low priority. Other editors, like Helix, need tree-sitter grammar and this one also needs to be written and actually maintained unlike all existing ones 😅

**Comment by mkaput** [comment]
Changing plans, thank y'all for feedback on DMs. I'm gonna prioritize this

**Comment by mkaput** [comment]
Lions work has been done and should be released in Cairo 2.7. I strongly suggest battle-testing latest nightlies and report any bugs 😍 

Left less important stuff has been moved to separate issues: #5823 #5824 #5826
